server/aesdsocket.c

```diff
@@ -49,130 +86,489 @@ void print_buf(char* buffer, int buffer_len)
+            close(parameters->client_fd);
+            pthread_exit(NULL);
+        }
+        printf("Can read from the client\r\n");
+        printf(" read from client buf length is %d\r\n", buffer_len);
+
+        if (buffer_len == 0)
+        {
+            printf("Disconnected from client\n");
+            break;
+        }
+
+        // stop with newline character
+        for (index = 0; index < buffer_len; index++)
+        {
+            if (buffer[index + len_tracker - MAX_SIZE] == '\n')
+            {
+                printf("New line detected\r\n");
+                break;
+            }
+        }
+
+        // if(buffer_len < MAX_SIZE)
+        if (index < buffer_len)
+        {
+            // can write to file directly
+            // fwrite(buffer, buffer_len, 1, fp);
+            fseek(fp, 0, SEEK_END);
+            mutex_lock = pthread_mutex_lock(parameters->mutex);
+            if (mutex_lock != 0)
+            {
+                printf("Error in acquring the mutex\r\n");
+                // Log this info
+                syslog(LOG_INFO, "Error in acquring the mutex. Error: %d\r\n", errno);
+                if(buffer)
+                {
+                free(buffer);
+                }
+                buffer = NULL;
+                graceful_exit();
+                close(parameters->client_fd);
+                pthread_exit(NULL);
+            }
+            printf("The mutex is locked\r\n");
+
+            if (fwrite(buffer, len_tracker - MAX_SIZE + index + 1, 1, fp) == -1)
+            {
+                printf("Error in writing to the file\r\n");
+                // Log this info
+                syslog(LOG_INFO, "Error in writing to the file. Error: %d\r\n", errno);
+                if(buffer != NULL)
+                {
+                free(buffer);
+                }
+                buffer = NULL;
+                graceful_exit();
+                close(parameters->client_fd);
+                pthread_exit(NULL);
+            }
+
+                printf("buf from fwrite to a file from buffer: ");
+                for (int i = 0; i < (len_tracker - MAX_SIZE + index + 1); i++)
+                    printf("%c", buffer[i]);
+                 
+        }
+        else
+        {
+                printf("Incrementing the buffer\r\n");
+                len_tracker += MAX_SIZE;
+                buffer = (char *)realloc(buffer, len_tracker);
+                if (buffer == NULL)
+                {
+                printf("Error in Malloc\r\n");
+                // Log this info
+                syslog(LOG_INFO, "Error in Malloc\r\n");
+                // thread_complete true as write is complete now
+                parameters->thread_complete = true;
+                }
+                printf("Incremented the buffer\r\n");
+        }
+        printf("Before fseek\r\n");
+        fseek(fp, 0, SEEK_SET);
+        printf("After fseek\r\n");
+
+        mutex_lock = pthread_mutex_unlock(parameters->mutex);
```

Comment:
@adna4700 I don't think your mutex unlock is in the correct place here, it will happen even if the mutex is not locked in the `if (index < buffer_len)` case, this will fail and then will exit.
You should check to see if this might be related to your issue when running on qemu by looking at your `/var/log/messages` log file.